Add documentation on how to do unit tests with web3.py

relyt29 opened the pull request:
### What was wrong?

The python ecosystem in general is sorely in need of updated documentation of how to do unit testing of solidity contracts with python. Something along the lines of: https://github.com/ethereum/pyethereum/wiki/Using-pyethereum.tester

### How was it fixed?

I wrote said example documentation.

#### Cute Animal Picture

![](https://i.imgflip.com/2ycipp.jpg)

relyt29 commented:
happy to rewrite this in pytest, give me 10 minutes

pipermerriam commented:
might be nice to actually write this as a file under the `./tests` directory and use the [include directives](http://docutils.sourceforge.net/docs/ref/rst/directives.html#include) to inline the file where you want it displayed.  This should ensure that this *example* test always actually works if someone wishes to copy/paste it from the docs.

lorenzb commented:
w00t

One suggestion: it might be useful to log an event from the contract and check its correctness from the example test.

relyt29 commented:
right, I intended to do that originally when I added the event and then forgot about it, thanks

relyt29 commented:
Okay I think this should be good for another review, I added code to check against an emitted event, and moved the code into the tests to be run with the test suite. Apologies if I put the file in the wrong place, I did my best guess and put it in the core/contracts directory.

pipermerriam commented:
Will try to go over this and get it merged monday.  Thanks for writing this up.

kclowes commented on docs/examples.rst:
```diff
@@ -449,3 +449,25 @@ When someone has an allowance they can transfer those tokens using the
+Contract Unit Tests in Python
+-----------------------------
+
+Here is an example of how one can use the pytest platform in python,web3.py,
```
missing a space here: `python, web3.py`

kclowes commented on docs/examples.rst:
```diff
@@ -449,3 +449,25 @@ When someone has an allowance they can transfer those tokens using the
+dependencies you can use the pinned extra for eth_tester in web3 and pytest:
+
+.. code-block:: bash
+    pip install web3[tester] pytest
```
It's convention in the docs to put a `$` in front of code that gets run from the terminal, so that should be added here. You'll also need a blank line in between `.. code-block:: bash` and `pip install web3[tester] pytest` to get it to show up as a code block.

kclowes commented on tests/core/contracts/test_contract_example.py:
```diff
@@ -0,0 +1,97 @@
+
+def test_set_bar(eth_tester, w3, deploy_contract):
+    contract_object = deploy_contract
+    account1 = str(eth_tester.get_accounts()[1])
```
nitpick: You don't need this `str` here either

kclowes commented on tests/core/contracts/test_contract_example.py:
```diff
@@ -0,0 +1,97 @@
+    #
+    # }
+
+    deploy_address = str(eth_tester.get_accounts()[0])
```
nitpick: you don't need this `str` here

kclowes commented on docs/examples.rst:
```diff
@@ -449,3 +449,25 @@ When someone has an allowance they can transfer those tokens using the
+
+.. include::  ../tests/core/contracts/test_contract_example.py
+    :code: python
+    :start_line: 1
```
If you change this to `start-line` the include should work correctly.

relyt29 commented:
Okay, made changes, thanks for the helpful hint on how to render the documentation locally - I think this should be good for another review

pipermerriam commented:
@kclowes I pushed a commit to this updating the test structure a bit (mostly just personal preference).  I'll leave it to you to do final review + merge.

kclowes commented:
Sweet! 🚀 Thanks @relyt29 and @pipermerriam!
